### ISSUE TITLE
Getting data as stale true for some filters in query

### ISSUE THREAD

**Issue by Manzurkds** [apollographql/apollo-client#4581]
I'm facing a very weird issue here.

We have a graphql search query, that takes in filters and gives us a  list of items.
The query is something like this

`allOpportunityV2(only: String, page: Int, per_page: Int, q: String, with: String, sort: String, filters: OpportunityFilter): OpportunityList`

OpportunityFilter is an object having filter selections like 
```
{
duration: RangeInput,
home_mcs: [Int],
programmes: [Int]
}
```

When I make this query, for certain filter combination, apollo is making a network request even if it was queried before and I'm getting proper response in the network tab, but in apollo's success method, I'm not getting the proper response.

**Intended outcome:**
Get a proper response when query is successful

**Actual outcome:**
```
{
data: undefined
loading: false
networkStatus: 7
stale: true
}
```

**How to reproduce the issue:**
It just happens randomly for some combination and is quite unpredictable to accurately reproduce. 

**Versions**
System:
    OS: macOS 10.14.3
  Binaries:
    Node: 10.15.0 - ~/.nvm/versions/node/v10.15.0/bin/node
    Yarn: 1.7.0 - /usr/local/bin/yarn
    npm: 6.4.1 - ~/.nvm/versions/node/v10.15.0/bin/npm
  Browsers:
    Chrome: 72.0.3626.121
    Firefox: 63.0.3
    Safari: 12.0.3
  npmPackages:
    apollo-angular: ^1.5.0 => 1.5.0
    apollo-angular-link-http: ^1.4.0 => 1.4.0
    apollo-cache-inmemory: ^1.4.2 => 1.4.2
    apollo-client: ^2.4.12 => 2.4.12
    apollo-link: ^1.2.6 => 1.2.6
    apollo-link-retry: ^2.2.8 => 2.2.8
    apollo-link-state: ^0.4.2 => 0.4.2

**Comment by Manzurkds** [comment]
I dug further into this issue and found out that it's because of the difference in the state of the variable passed to apollo.

So I have this variable let's say `queryVariables` and I reference one of its key called `filters` to my application's `this.filters` which for graphql is of type `OpportunityFilter`

Now at a certain place, I'm modifying `this.filters` value which is not in accordance of `OpportunityFilter`, but just before calling apollo query, I'm reassigning `this.filters` with new values which are in accordance of the `OpportunityFilter` like this

```
this.filters = this.makeFilters() // Makes a fresh object

// At this point of time, queryVariables is with modified filters, which is not what I wanted.
// Lets call this point1

this.queryVariables = {
	page: 1,
	per_page: this.perPageLimit,
	q: this.searchInput,
	filters: this.filters
	sort: this.selections.sort
	};
// At this point of time, queryVariables is with proper filters, which is what I wanted
// Lets call this point2

this.makeApolloQuery(this.query, this.queryVariables);
// But apollo treats this.queryVariables like it is in point1 instead of point2. That's where the problem occurs
```
Since the variable at point1 isn't valid, hence it isn't associated with any query, hence, the it is not able to get the data from store and makes a network request, and even then for some unknown reason I get data as undefined.


In order to overcome this issue, I thought of a workaround where I will reset the cache, if something like this occurs, but while doing that, when I was reading a query with the readQuery method on apollo's cache, I understand what's going on.
So I tried to debug this issue and saw that, my variables are being passed correctly (like in point2), but it's treated the same way as in point1.
What's even funny is if at that point of time, I just copy my variable and reassign it to a new variable, and call the readQuery method, it works!!
Going even deeper I found that in cache.js, there's a method called `cache.get(key)` when I was debugging, the key was set as `{}` and it was throwing me the error which I was receiving. But interestingly here too, if I pass in `{}` instead of `key` to the method, like `cache.get({})` which essentially is the same thing, I get a different response, this time I get `undefined`.

I've got a workaround for this issue, by using spread operator while assigning the filters key in variables, but still curious as though why is this happening and it'll really be helpful if this is fixed.


I've attached a video demonstrating the effect variables have, by called `read` method in apollo's library itself.
https://drive.google.com/open?id=1ePN7SLehf8KFkOAfWGizrf8jB6WIw5pM

Thank you!
